management-service/models/custom_segmenter.go

```diff
@@ -0,0 +1,678 @@
+		}
+		valueSet.Insert(value)
+	}
+	return nil
+}
+
+// ValidateSegmenterNotPreRequisiteOfItself checks if all the segmenter names of all the prerequisites specified for all the
+// constraints in the segmenter do not refer to the actual segmenter's name.
+func (s *CustomSegmenter) ValidateSegmenterNotPreRequisiteOfItself() error {
+	if s.Constraints == nil {
+		return nil
+	}
+
+	for _, constraint := range *s.Constraints {
+		for _, preRequisite := range constraint.PreRequisites {
+			if preRequisite.SegmenterName == s.Name {
+				return fmt.Errorf("segmenter %s cannot be a prerequisite of itself", preRequisite.SegmenterName)
+			}
+		}
+	}
+	return nil
+}
+
+// ValidateConstraintValues checks if all the allowed values and options of all the constraints in the segmenter are
+// present within the options specified. If no options or no constraints are specified, this check passes.
+func (s *CustomSegmenter) ValidateConstraintValues() error {
+	if s.Options == nil || s.Constraints == nil {
+		return nil
+	}
+	var optionsSlice []interface{}
+	for _, value := range *s.Options {
+		optionsSlice = append(optionsSlice, value)
+	}
+	segmenterValueSet := set.New(optionsSlice...)
+
+	for _, constraint := range *s.Constraints {
+		// checks if all the allowed values are present in the segmenter options
+		if len(constraint.AllowedValues) == 0 {
+			return fmt.Errorf("allowed values cannot be an empty array")
+		}
+		for _, allowedValue := range constraint.AllowedValues {
+			if !segmenterValueSet.Has(allowedValue) {
+				return fmt.Errorf("allowed value %s is not specified within segmenter options", allowedValue)
+			}
+		}
+
+		// checks if all the constraint options values specified are present in the segmenter options and that they
+		// match every value in the allowed values field
+		if constraint.Options != nil {
+			err := validateOptionsHaveUniqueValues(constraint.Options)
+			if err != nil {
+				return err
+			}
+
+			var constraintOptionsSlice []interface{}
+			for constraintName, constraintValue := range *constraint.Options {
+				if !segmenterValueSet.Has(constraintValue) {
+					return fmt.Errorf("segmenter name %s with value %s is not specified within segmenter options",
+						constraintName, constraintValue)
+				}
+				constraintOptionsSlice = append(constraintOptionsSlice, constraintValue)
+			}
+
+			allowedValuesSet := set.New(constraint.AllowedValues...)
+			constraintOptionsSet := set.New(constraintOptionsSlice...)
+
+			if constraintOptionsSet.Difference(allowedValuesSet).Len() != 0 ||
+				allowedValuesSet.Difference(constraintOptionsSet).Len() != 0 {
+				return fmt.Errorf("segmenter values in constraint options do not match those in the allowed values")
+			}
+		}
+	}
+	return nil
+}
+
+// ConvertToTypedValues converts a CustomSegmenter's segmenter values that are untyped to the type specified in
+// the Type field. This method is also used to validate the unknown segmenter value types (i.e. validate values passed
```

Comment:
> This method is also used to validate the unknown segmenter value types (i.e. validate values passed
// in as user input with respect to the specified type)

Not sure I follow this comment entirely. Do you mean to say that since we will run the type casting on each value, it will be indirectly validating the values?

Reply:
Yep that's right; the main purpose of the function is to perform the type cast based on the `type` of the segmenter, but doing that indirectly validates the values too.

Reply:
Got it. Can it be rephrased to include "indirectly validates the type of each value"? I feel that would be clearer.

Reply:
Okay I've added that phrase in, and now it looks like: 
> ConvertToTypedValues converts a CustomSegmenter's segmenter values that are untyped to the type specified in  the Type field. As this method also indirectly validates the type of each value, it is also used to validate unknown segmenter value types (i.e. validate values passed in as user input with respect to the specified type)